internal/tools/rootfs/main.go

```diff
@@ -0,0 +1,111 @@
+// --------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// --------------------------------------------------------------------------------------------
+
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/ext4/tar2ext4"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/daemon"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+)
+
+func main() {
+	image := flag.String("i", "", "image")
+	tag := flag.String("t", "latest", "tag")
+	destination := flag.String("d", "local", "destination")
+
+	flag.Parse()
+	if flag.NArg() != 0 || len(*image) == 0 {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	layers, err := getRootfsLayerHashes(*image, *tag, *destination)
+
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+
+	for i, layer := range layers {
+		fmt.Printf("[%d]: %s\n", i, layer)
+	}
+}
+
+func getRootfsLayerHashes(image_name string, image_tag string, destination string) ([]string, error) {
+	hashes := []string{}
+	image_string := image_name + ":" + image_tag
+
+	if len(image_name) == 0 || len(image_tag) == 0 {
+		return hashes, fmt.Errorf("'%s' is not a valid image name", image_string)
+	}
+
+	validDestinations := map[string]bool{
+		"local":  true,
+		"remote": true,
+	}
+	if !validDestinations[destination] {
+		fmt.Print("Destination value should fall in following values:[")
+		for k := range validDestinations {
+			fmt.Print(k + ",")
+		}
+		fmt.Println("]")
+		return hashes, errors.New("invalid destination")
+	}
+
+	ref, err := name.ParseReference(image_string)
+	if err != nil {
+		return hashes, fmt.Errorf("'%s' is not a valid image name", image_string)
+	}
+
+	// by default, using local as destination
+	var localImageOption []daemon.Option
+	image, err := daemon.Image(ref, localImageOption...)
+	if strings.ToLower(destination) == "remote" {
+		var imageOptions []remote.Option
+		image, err = remote.Image(ref, imageOptions...)
+	}
+	if err != nil {
+		return hashes, fmt.Errorf("unable to fetch image '%s': %s", image_string, err.Error())
+	}
+
+	layers, err := image.Layers()
+	if err != nil {
+		return hashes, err
+	}
+
+	for _, layer := range layers {
```

Comment:
Lets declare `hashes` above this loop (`var hashes []string`, or however you prefer) and just `return nil, err` everywhere else and not `return hashes, err`. It's a lot easier to read if the declaration is right above where we actually make use of it/fill the slice up.